### ISSUE TITLE
[RAD] I'm completely lost...

### ISSUE THREAD

**Issue by rozek** [amark/gun#1329]
The title says everything: I'm completely lost...

I added RAD to my browser as mentioned in the [docs](https://gun.eco/docs/RAD) and continued evaluating GUN - until I saw that there were only two keys in my "IndexedDB": namely "%1C" and "!"

So I started evaluating RAD using a `trivialRADStore`:

```
  const trivialRADStore_Store = Object.create(null)

  const trivialRADStore = {
    get: function (Key, Callback) {
      console.log('  trivialRADStore: reading', Key)
      return Callback(null,trivialRADStore_Store['' + Key])
    },

    put: function (Key, Data, Callback) {
      console.log('  trivialRADStore: writing', Key, '(' + typeof Data + ')')
      trivialRADStore_Store['' + Key] = Data
      return Callback(null,1)
    }
  }
```

which I could directly write to:

```
  const Rad = window.Radisk({ store:trivialRADStore })

  async function RAD_read (Key, Options) {
    return new Promise((resolve,reject) => {
      Rad(Key, (Error, Data, Info) => {
        console.log('  RAD Info:', Info)
        return (
          Error == null ? resolve(Data) : reject(Error)
        )
      }, Options)
    })
  }

  async function RAD_write (Key, Data) {
    return new Promise((resolve,reject) => {
      Rad(Key, Data, (Error, Data) => {
        return (
          Error == null ? resolve(Data) : reject(Error)
        )
      })
    })
  }
```

Surprisingly, my RAD adapter was also called with two keys only (the same as above). (However, I also logged the "info" record returned by the "read" variant of the `RAD` function - and that sometimes reported it would have parsed "1048576" bytes/chars although I had not written much at that point....)

Then I had the idea that I might have written too few keys and too little data - hence I wrote a simple function to fill my store

```
  async function writeNested (BaseKey,Depth) {
    for (let i = 0; i < 10; i++) {
      const currentKey = (BaseKey === '' ? '' : BaseKey + '/') + i
      await RAD_write(currentKey,currentKey)

      if (Depth > 1) { await writeNested(currentKey,Depth-1) }
    }
  }
```

and that soon started to take looooong to write a new entry - and had to be stopped after very few write calls.

So: what am I doing wrong? Where is the "radix tree" promised by the docs?

**Comment by rozek** [comment]
a small note: if I change `RAD_write` and set the `trivialRADStore_Store` directly (i.e., avoid to call `Rad`), the whole loop completes almost immediately - thus, its `RAD` that cause the problems...

**Comment by rozek** [comment]
Ok, my current observations so far:

* entry "!" seems to collect written entries until they get written at their own keys
* I have no idea what entry "%1C" is good for - indeed I had to suppress reading and writing that entry!
* keys get encoded in a strange way: "1/1" becomes "1%2F1"
* there is no "radix tree" anywhere
* entries are stored as JSON strings - but in a really strange format: writing "1/1" to key "1/1" (yes, in my experiments I used the key as value as well) creates an entry at "1%2F1" with the contents {"1/1":{"":"1/1"}}

In order to speed my in-memory(!) store up, I had to change my RAD adpter's code to

```
  const trivialRADStore_Store = Object.create(null)

  const trivialRADStore = {
    get: function (Key, Callback) {
      console.log('  trivialRADStore: reading', Key)
if (Key === '%1C') { return Callback(null,undefined) }
      return Callback(null,trivialRADStore_Store['' + Key])
    },

    put: function (Key, Data, Callback) {
      console.log('  trivialRADStore: writing', Key, typeof Data)
if (Key === '%1C') { return Callback(null,0) }
      trivialRADStore_Store['' + Key] = Data
      return Callback(null,1)
    }
  }
```

This is nothing but a hack - however, without that hack, my browser starts slowing down heavily after a few writes...

**Comment by rozek** [comment]
RAD also strangely reads the previously written key again - and writes it out again, unchanged! How useless is that?

**Comment by rozek** [comment]
here is a transcript of the first few operations on my in-memory store:

```
entry #1: 0
   trivialRADStore: reading %1C 0 undefined
   trivialRADStore: reading ! 0 undefined
   trivialRADStore: writing %1C 12 {"!":{"":1}}
   trivialRADStore: writing ! 20 {"0":{"":"value 0"}}
 entry #2: 0/0
   trivialRADStore: reading ! 20 {"0":{"":"value 0"}}
   trivialRADStore: writing %1C 25 {"!":{"":1},"0/0":{"":1}}
   trivialRADStore: writing 0%2F0 24 {"0/0":{"":"value 0/0"}}
   trivialRADStore: writing ! 20 {"0":{"":"value 0"}}
 entry #3: 0/1
   trivialRADStore: reading 0%2F0 24 {"0/0":{"":"value 0/0"}}
   trivialRADStore: writing %1C 41 {"!":{"":1},"0/":{"0":{"":1},"1":{"":1}}}
   trivialRADStore: writing 0%2F1 24 {"0/1":{"":"value 0/1"}}
   trivialRADStore: writing 0%2F0 24 {"0/0":{"":"value 0/0"}}
 entry #4: 0/2
   trivialRADStore: reading 0%2F1 24 {"0/1":{"":"value 0/1"}}
   trivialRADStore: writing %1C 52 {"!":{"":1},"0/":{"0":{"":1},"1":{"":1},"2":{"":1}}}
   trivialRADStore: writing 0%2F2 24 {"0/2":{"":"value 0/2"}}
   trivialRADStore: writing 0%2F1 24 {"0/1":{"":"value 0/1"}}
```

**Comment by rozek** [comment]
well, entry "%1C" seems to contain, what is called the "radix tree": the JSON string of an object with the hierarchy of keys.

I have no idea what that is good for as the get and put calls to my adapter always contain the full key of the affected entry...

**Comment by amark** [comment]
RAD being replaced with Book, just heads up. So if you're willing to help hack on that instead, please let me know! We do screen calls. @rogowski has been helping.

RAD chunks the tree into pages, so what you're seeing is the pages name and the root file that keeps track of the folder of files. More explained here:

https://youtu.be/5fCPRY-9hkc?t=1275

but note: Book will be replacing this, so don't spend too much time/effort on it.

**Comment by rozek** [comment]
That's good to know - I already lost a day on this because I searched the problem on my side...will give up on RAD then!

**Comment by jadbox** [comment]
@amark Where can I find information on Book?

**Comment by amark** [comment]
@rozek I'm back to working on Book, was distracted with [@SecureRender](https://twitter.com/securerender) for a while. Want to help me & Rogowski?

@jadbox same??? Sorry not much docs yet, mostly just ask me on http://chat.gun.eco and maybe look thru tests in tests/rad/book (or... somewhere like that) or jump on a call with me (DM on twitter).

**Comment by k98kurz** [comment]
> [@rozek](https://github.com/rozek) I'm back to working on Book, was distracted with [@SecureRender](https://twitter.com/securerender) for a while. Want to help me & Rogowski?
> 
> [@jadbox](https://github.com/jadbox) same??? Sorry not much docs yet, mostly just ask me on http://chat.gun.eco and maybe look thru tests in tests/rad/book (or... somewhere like that) or jump on a call with me (DM on twitter).

So is gundb fundamentally broken at the moment, then? Is localStorage the only semi-reliable option?

**Comment by mblaney** [comment]
> > [@rozek](https://github.com/rozek) I'm back to working on Book, was distracted with [@SecureRender](https://twitter.com/securerender) for a while. Want to help me & Rogowski?
> > [@jadbox](https://github.com/jadbox) same??? Sorry not much docs yet, mostly just ask me on http://chat.gun.eco and maybe look thru tests in tests/rad/book (or... somewhere like that) or jump on a call with me (DM on twitter).
> 
> So is gundb fundamentally broken at the moment, then? Is localStorage the only semi-reliable option?

@k98kurz yes rad is broken in gundb, I tried fixing it in the current version but ended up working from the much earlier https://github.com/gundb/port which I've since forked to https://github.com/mblaney/holster

**Comment by k98kurz** [comment]
> > So is gundb fundamentally broken at the moment, then? Is localStorage the only semi-reliable option?
> 
> [@k98kurz](https://github.com/k98kurz) yes rad is broken in gundb, I tried fixing it in the current version but ended up working from the much earlier https://github.com/gundb/port which I've since forked to https://github.com/mblaney/holster

That is pretty disappointing. I spent over a hundred hours building a markdown editor app using gundb as the storage/sync system and SEA for encryption, and it just yesterday stopped working because localStorage refuses to accept writes after just 200 kb stored in the `gun/` key (can't even delete old documents to make room). Now I can't get IndexedDB working, so it feels like the whole thing has been a waste of time and effort that I could have instead spent working on my own CRDT synchronization system. Perhaps the gundb marketing material/documentation should note this fundamental flaw instead of misleading devs into thinking this system works properly.

**Comment by mblaney** [comment]
I did exactly the same thing, hundreds of hours building an app which then only stopped working with gun once storage reached a size where the radisk bugs became a problem. I've migrated the app to holster and it's working fine now.

**Comment by k98kurz** [comment]
> I did exactly the same thing, hundreds of hours building an app which then only stopped working with gun once storage reached a size where the radisk bugs became a problem. I've migrated the app to holster and it's working fine now.

Hmm. I'll give it a shot in the near future.
